### PR TITLE
started design vignette

### PR DESCRIPTION
**Pull request by sckott**
Design chapter, via https://github.com/ropensci-books/http-testing/issues/30

Probably some other things could be included, but I think this is a good first pass

#232

### CODE REVIEW

File: man/rmdhunks/vcr-design.Rmd [hunk 1]
```diff
@@ -0,0 +1,152 @@
+## Design
+
+This section explains `vcr`'s internal design and architecture.
+
+## Where vcr comes from and why R6
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:5]: Link to something explaining R6? https://adv-r.hadley.nz/r6.html?

File: man/rmdhunks/vcr-design.Rmd [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+As described above, `vcr` uses R6 internally, but users interact with
+normal R functions. 
+
+```r
+functions == business
+R6 == party
+```
+
+The business and party could be reversed, depending on your perspective.
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:29]: I did not understand this image right away. Add a sentence about the hairstyle?

File: man/rmdhunks/vcr-design.Rmd [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+used in `vcr`. 
+
+### Class/function names are inherited from Ruby vcr
+
+Since R `vcr` was ported from Ruby, we kept most of the names of 
+functions/classes and variables. So if you're wondering about why 
+a function, class, or variable has a particular name, its derivation 
+can not be found out in this package, for the most part that is.
+
+### Hooks into http clients
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:40]:
```suggestion
### Hooks into HTTP clients
```

File: man/rmdhunks/vcr-design.Rmd [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+vcr package literally overrides certain methods in Ruby HTTP clients, 
+hijacking internals of the HTTP clients. 
+
+However, monkey patching is not allowed in R. Thus, in R we have to
+somehow have "hooks" into HTTP clients in R. Fortunately, Scott is the
+maintainer of one of the HTTP clients, `crul`, so was able to quickly
+create a hook. Very fortunately, there was already a hook mechanism
+in the `httr` package. 
+
+The actual hooks are not in `vcr`, but in `webmockr`. `vcr` depends on
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:57]: a blog post about how the hooks work (that could then get integrated into webmockr docs) would be great!

File: man/rmdhunks/vcr-design.Rmd [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+a few methods in it, so there's a lot going on in the class. Ideally
+the class would be separated into subclasses to handle similar sets
+of logic, but there's not an easy way to do that with R6. 
+
+Of note in `Cassette` is that when called, within the `initialize()` 
+call `webmockr` is used to create webmockr stubs. 
+
+### How HTTP requests are handled
+
+Within `webmockr`, there's calls to the vcr class `RequestHandler`, which
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:87]:
```suggestion
Within `webmockr`, there are calls to the vcr class `RequestHandler`, which
```

File: man/rmdhunks/vcr-design.Rmd [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+with no real HTTP request allowed.
+- **Recordable** This is an HTTP request for which no match is found
+in the cassette defined in the `use_cassette()`/`insert_cassette()` call.
+In this case a real HTTP request is allowed, and the request/response is
+recorded to the cassette.
+- **Unhandled** This is a group of cases, all of which cause an error
+to be thrown with a message trying to help the user figure out how to
+fix the problem.
+
+If you use vcr logging you'll see these categories in your logs.
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:106]: example of a log?

File: man/rmdhunks/vcr-design.Rmd [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+### Serializers
+
+Serializers handle in what format cassettes are written to files on disk.
+The current options are YAML (default) and JSON. 
+
+An R6 class `Serializer` is the parent class for all serializer types;
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:113]: Maybe add context on why you implemented YAML first and the problems of YAML?

File: man/rmdhunks/vcr-design.Rmd [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+### Logging
+
+An internal environment (`vcr_log_env`) is used when you use logging.
+At this point it only keeps track of one variable - `file` - to be able
+to refer to what file is used for logging across many classes/functions
+that need to write to the log file.
+
+### A bit of housekeeping
+
+Another internal environment (`vcr__env`) is used to keep track of a
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:131]: why two "_"?

File: man/rmdhunks/vcr-design.Rmd [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+few items, including the current cassette in use, and the last vcr error.
+
+### Lightswitch
+
+Another internal environment (`light_switch`) is used to keep track of users
+turning on and off `vcr`. See `?lightswitch`.
+
+
+
+## Footnotes
```
Comment (maelle) [inline man/rmdhunks/vcr-design.Rmd:141]: given that `bookdown::bs4_book()` makes footnotes accessible on hover, and that pkgdown next version will do that as well, I think you should remove this section heading.

**Review by maelle — COMMENTED** [review]
Awesome, super interesting! I added a few comments.

Once this is merged (not urgent) could you make a PR to the book repo?

File: man/rmdhunks/vcr-design.Rmd (hunk 1 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:5]: yes probably that

File: man/rmdhunks/vcr-design.Rmd (hunk 2 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:29]: AH, ok, not a universal thing, might change that

File: man/rmdhunks/vcr-design.Rmd (hunk 4 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:57]: good idea, probably won't have time for that though

File: man/rmdhunks/vcr-design.Rmd (hunk 6 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:106]: probably not useful to show a log file, but can link to help docs on logging

File: man/rmdhunks/vcr-design.Rmd (hunk 7 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:113]: good idea

File: man/rmdhunks/vcr-design.Rmd (hunk 8 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:131]: i dont remember :)

File: man/rmdhunks/vcr-design.Rmd (hunk 9 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:141]: will do

File: man/rmdhunks/vcr-design.Rmd (hunk 1 above)
Comment (sckott) [reply man/rmdhunks/vcr-design.Rmd:5]: aldready linking to it in the paragraph

**Pull request merged**